CMake minor fixes

jngrad opened the pull request:
Description of changes:
- change next milestone to 4.2
- load `GNUInstallDirs` to make standard GNU paths accessible from CMake variables
- simplify CMake logic and install in `python3.X` folder instead of the deprecated `python3` folder
- add extra check to make sure install paths are correctly configured (all python and shared object files must be inside the package `espressomd`)

jngrad committed:
Next version is 4.2

jngrad committed:
Load standard GNU variables in CMake

jngrad committed:
Add extra CMake install path check

fweik commented on CMakeLists.txt:
```diff
@@ -20,6 +20,7 @@
 
 cmake_minimum_required(VERSION 3.4)
 include(FeatureSummary)
+include(GNUInstallDirs)
```
Probably the block
```cmake
if(NOT DEFINED BINDIR)
  set(BINDIR "bin")
endif(NOT DEFINED BINDIR)
```
in line 401 should be removed now. Did you check that the paths provided by `GNUInstallDirs` are used for all install targets? I did a quick grep and did find nothing, but I only checked the `install` commands, not sure
if this effects rpaths and so on....

jngrad replied:
Thanks! I overlooked that one. The `install()` commands have to be configured such that submodules get installed in the python `espressomd` folder, if not the new test in c43d057 will detect that (it searches for misplaced files in `CMAKE_INSTALL_PREFIX`). The GNU variables inherit from `CMAKE_INSTALL_PREFIX`, so we should use them as often as possible.

jngrad committed:
Use GNU variable BINDIR

jngrad commented:
I've a few more improvements.

jngrad committed:
Remove obsolete Python guard

The Python guard isn't necessary anymore because submodules are
now installed in the Python package espressomd, and the testsuite
itself already has a Python guard in the top-level CMakeLists.txt.

jngrad commented:
The failures in CI are actually a bug in the build system, not a regression. Working on it.

jngrad committed:
Use modern CMake install procedure

Remove the non-standard CMAKE_INSTALL_EXEC_PREFIX variable in
favor of CMAKE_INSTALL_PREFIX. Install espressomd in the standard
python3.X folder instead of the deprecated python3 folder. Remove
duplicated code logic. Use explicit keyword arguments in calls to
functions from distutils.sysconfig.

jngrad commented on CMakeLists.txt:
```diff
@@ -241,7 +242,7 @@ if(WITH_PYTHON)
       COMMAND
         ${PYTHON_EXECUTABLE} -c
-        "import distutils.sysconfig as cg; print(cg.get_python_lib(1,0,prefix='${CMAKE_INSTALL_EXEC_PREFIX}'))"
+        "import distutils.sysconfig as cg; print(cg.get_python_lib(prefix='${CMAKE_INSTALL_PREFIX}', plat_specific=True, standard_lib=False).replace('${CMAKE_INSTALL_PREFIX}/', '', 1))"
```
@junghans this change might affect your packaging workflow

junghans replied:
Thanks for asking, but it won't as `PYTHON_INSTDIR` is set for packaging, so this call isn't actually used.

jngrad commented:
bors r=fweik
